core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -297,10 +418,16 @@ class LogSegment(val log: FileMessageSet,
   def lastModified = log.file.lastModified
 
   /**
+   * The largest timestamp this segment contains.
+   */
+  def largestTimestamp = if (maxTimestampSoFar >= 0) maxTimestampSoFar else lastModified
```

Comment:
Currently, we have the logic to force adding an index entry (with last modified timestamp) if none of the message has timestamp. I am wondering why this is necessary. It seems that we can just leave the timeindex empty in that case and largestTimestamp() can still return lastModified? That may make the log recovery logic a bit simpler.